feat(ffi): Add FFI bindings to `Room::forget`

jmartinesp opened the pull request:
This PR adds bindings for `Room::forget` and makes sure you can also forget rooms with `Banned` state.

- [ ] Public API changes documented in changelogs (optional)

<!-- Sign-off, if not part of the commits -->
<!-- See CONTRIBUTING.md if you don't know what this is -->
Signed-off-by:

bnjbvr commented on crates/matrix-sdk/src/client/mod.rs:
```diff
@@ -2471,6 +2471,12 @@ impl Client {
         Ok(Room::new(self.clone(), base_room))
     }
+
+    /// Remove the room from the store, removing all data related to it.
+    pub async fn forget_room(&self, room_id: &RoomId) -> Result<()> {
+        self.inner.base_client.forget_room(room_id).await?;
+        Ok(())
+    }
```
There's `Room::forget()`, and it would be easy to mistake one for the other. If you plan to keep this one, now you know there's `Room::forget()`: can you precise in the doc comment this does NOT forget the room at the server level, but only locally? (and as such, the room can still be in the DM list, etc.)

jmartinesp replied:
Oh, I didn't realise this was already there. I'll remove it and add bindings to `Room::forget` instead.

bnjbvr commented on crates/matrix-sdk/src/client/mod.rs:
```diff
@@ -3182,4 +3188,19 @@ pub(crate) mod tests {
+
+    #[async_test]
+    async fn test_forget_room() {
+        let server = MatrixMockServer::new().await;
+        let client = server.client_builder().build().await;
```
nit: you don't need the server here, you can create the client immediately

bnjbvr commented on bindings/matrix-sdk-ffi/src/client.rs:
```diff
@@ -1018,6 +1018,17 @@ impl Client {
     }
 
+    /// Leave a room by its room ID.
+    pub async fn leave_room_by_id(
```
What's the purpose of this function, when you have `Room::leave()`? In other words, what are the conditions that make it necessary: a room that's out of sync shouldn't be present at all, and having "bad state locally" seems like a poor justification for API proliferation.

jmartinesp replied:
I think you're right and we should focus on fixing the underlying issue in `RoomPreview` instead.

jmartinesp commented:
I removed the `leave_room_by_id` function and moved the forget room bindings so they match the existing API, including the API call to the `/forget` endpoint.

bnjbvr commented on crates/matrix-sdk/src/room/mod.rs:
```diff
@@ -2896,11 +2896,11 @@ impl Room {
         let state = self.state();
-        if state != RoomState::Left {
-            return Err(Error::WrongRoomState(WrongRoomState::new("Left", state)));
+        if matches!(state, RoomState::Left | RoomState::Banned).not() {
```
Can you use an explicit `match` here? Next time we add a `RoomState` variant, the compiler will complain that we need to handle it, which is nicer than having to remember to handle it ourselves.

bnjbvr commented on crates/matrix-sdk/tests/integration/room/left.rs:
```diff
@@ -77,6 +77,67 @@ async fn test_forget_non_direct_room() {
+    event_cache.enable_storage().unwrap();
+
+    Mock::given(method("POST"))
+        .and(path_regex(r"^/_matrix/client/r0/rooms/.*/forget$"))
```
Bonus points if you add new endpoints to the `MatrixMockServer` :eyes:

bnjbvr reviewed (approved):
Thanks!
